File: src/main/java/com/signature/hdlobf/HashFunctions.java
package com.signature.hdlobf;

public class HashFunctions {
    public static String hash1(String str) {
        int hash = 5381;
        for (int i = 0; i < str.length(); i++) {
            hash = ((hash << 5) + hash) + str.charAt(i);
        }
        long ret_hash = (hash & 0x7FFFFFFF);
        return Long.toHexString(ret_hash);
    }

    public static String hash2(String str) {
        int hash = 0;
        for (int i = 0; i < str.length(); i++) {
            if ((i & 1) == 0) {
                hash ^= ((hash << 7) ^ str.charAt(i) ^ (hash >> 3));
            } else {
                hash ^= (~((hash << 11) ^ str.charAt(i) ^ (hash >> 5)));
            }
        }
        long ret_hash = (hash & 0x7FFFFFFF);
        return Long.toHexString(ret_hash);
    }
}

File: src/hdl_obfuscator/mangler/HashFunctions.py
#!/bin/env python3

def hash1(string) -> str:
    hash_value = 5381
    for char in string:
        hash_value = ((hash_value << 5) + hash_value) + ord(char)
    ret_hash = hash_value & 0x7FFFFFFF
    return format(ret_hash, "08x")


def hash2(string) -> str:
    hash_value = 0
    for i, char in enumerate(string):
        if i % 2 == 0:
            hash_value ^= (hash_value << 7) ^ ord(char) ^ (hash_value >> 3)
        else:
            hash_value ^= ~((hash_value << 11) ^ ord(char) ^ (hash_value >> 5))
    ret_hash = hash_value & 0x7FFFFFFF
    return format(ret_hash, "08x")
